File: wave.py
# wave.py
import argparse
import math
import struct
import wave
import os
import subprocess
import platform
from typing import Optional

class WaveGenerator:
    def __init__(self, freq=440, amp=0.8, duration=2.0, sample_rate=44100, wave_type='sine'):
        self.freq = freq
        self.amp = amp
        self.duration = duration
        self.sample_rate = sample_rate
        self.wave_type = wave_type
        self.samples = []
        self._generate()

    def _generate(self):
        num_samples = int(self.sample_rate * self.duration)
        for i in range(num_samples):
            t = i / self.sample_rate
            phase = 2 * math.pi * self.freq * t
            if self.wave_type == 'sine':
                val = math.sin(phase)
            elif self.wave_type == 'square':
                val = 1 if math.sin(phase) >= 0 else -1
            elif self.wave_type == 'sawtooth':
                val = 2 * (phase / (2 * math.pi) - math.floor(phase / (2 * math.pi) + 0.5))
            elif self.wave_type == 'triangle':
                val = 2 * abs(2 * (phase / (2 * math.pi) - math.floor(phase / (2 * math.pi) + 0.5))) - 1
            elif self.wave_type == 'noise':
                import random
                val = random.uniform(-1, 1)
            else:
                val = 0
            self.samples.append(int(self.amp * val * 32767))

    def save_wav(self, filename: str):
        with wave.open(filename, 'w') as wf:
            wf.setnchannels(1)
            wf.setsampwidth(2)  # 16-bit
            wf.setframerate(self.sample_rate)
            data = struct.pack('<' + 'h' * len(self.samples), *self.samples)
            wf.writeframes(data)

    def play(self, filename: str):
        system = platform.system()
        try:
            if system == 'Windows':
                os.startfile(filename)
            elif system == 'Darwin':
                subprocess.run(['afplay', filename], check=True)
            else:
                players = ['ffplay', 'aplay', 'mpg123']
                for player in players:
                    try:
                        subprocess.run([player, filename], check=True)
                        return
                    except FileNotFoundError:
                        continue
                print('No suitable player found.')
        except Exception as e:
            print(f'Playback error: {e}')

def interactive():
    print('=== Waveform Audio Generator ===')
    types = ['sine', 'square', 'sawtooth', 'triangle', 'noise']
    wtype = input(f'Enter type ({", ".join(types)}, default: sine): ').strip() or 'sine'
    if wtype not in types:
        wtype = 'sine'
    freq = float(input('Enter frequency (Hz, default: 440): ').strip() or '440')
    amp = float(input('Enter amplitude (0.0-1.0, default: 0.8): ').strip() or '0.8')
    dur = float(input('Enter duration (seconds, default: 2.0): ').strip() or '2.0')
    rate = int(input('Enter sample rate (Hz, default: 44100): ').strip() or '44100')
    fname = input('Enter output filename (default: output.wav): ').strip() or 'output.wav'
    if not fname.endswith('.wav'):
        fname += '.wav'
    gen = WaveGenerator(freq, amp, dur, rate, wtype)
    gen.save_wav(fname)
    print(f'Saved to {fname}')
    if input('Play? (y/n, default: y): ').strip().lower() != 'n':
        gen.play(fname)

def cli():
    parser = argparse.ArgumentParser(description='Waveform Audio Generator')
    parser.add_argument('-f', '--freq', type=float, default=440, help='Frequency in Hz')
    parser.add_argument('-t', '--type', default='sine', choices=['sine','square','sawtooth','triangle','noise'], help='Waveform type')
    parser.add_argument('-a', '--amp', type=float, default=0.8, help='Amplitude (0.0-1.0)')
    parser.add_argument('-d', '--dur', type=float, default=2.0, help='Duration in seconds')
    parser.add_argument('-r', '--rate', type=int, default=44100, help='Sample rate')
    parser.add_argument('-o', '--output', default='output.wav', help='Output filename')
    parser.add_argument('--no-play', action='store_true', help='Don\'t play after generation')
    args = parser.parse_args()
    gen = WaveGenerator(args.freq, args.amp, args.dur, args.rate, args.type)
    gen.save_wav(args.output)
    print(f'Saved to {args.output}')
    if not args.no_play:
        gen.play(args.output)

if __name__ == '__main__':
    import sys
    if len(sys.argv) > 1:
        cli()
    else:
        interactive()

File: wave.js
// wave.js
const fs = require('fs');
const readline = require('readline');
const { exec } = require('child_process');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class WaveGenerator {
    constructor(freq = 440, amp = 0.8, duration = 2.0, sampleRate = 44100, waveType = 'sine') {
        this.freq = freq;
        this.amp = amp;
        this.duration = duration;
        this.sampleRate = sampleRate;
        this.waveType = waveType;
        this.samples = [];
        this.generate();
    }

    generate() {
        const numSamples = Math.floor(this.sampleRate * this.duration);
        for (let i = 0; i < numSamples; i++) {
            const t = i / this.sampleRate;
            const phase = 2 * Math.PI * this.freq * t;
            let val = 0;
            switch (this.waveType) {
                case 'sine':
                    val = Math.sin(phase);
                    break;
                case 'square':
                    val = Math.sin(phase) >= 0 ? 1 : -1;
                    break;
                case 'sawtooth':
                    val = 2 * (phase / (2 * Math.PI) - Math.floor(phase / (2 * Math.PI) + 0.5));
                    break;
                case 'triangle':
                    val = 2 * Math.abs(2 * (phase / (2 * Math.PI) - Math.floor(phase / (2 * Math.PI) + 0.5))) - 1;
                    break;
                case 'noise':
                    val = Math.random() * 2 - 1;
                    break;
                default:
                    val = 0;
            }
            this.samples.push(Math.round(this.amp * val * 32767));
        }
    }

    saveWAV(filename) {
        const dataSize = this.samples.length * 2;
        const byteRate = this.sampleRate * 2;
        const buffer = Buffer.alloc(44 + dataSize);
        let offset = 0;
        const writeU32 = (v) => { buffer.writeUInt32LE(v, offset); offset += 4; };
        const writeU16 = (v) => { buffer.writeUInt16LE(v, offset); offset += 2; };
        // RIFF
        buffer.write('RIFF', 0);
        writeU32(36 + dataSize);
        buffer.write('WAVE', 8);
        // fmt
        buffer.write('fmt ', 12);
        writeU32(16);
        writeU16(1);
        writeU16(1);
        writeU32(this.sampleRate);
        writeU32(byteRate);
        writeU16(2);
        writeU16(16);
        // data
        buffer.write('data', offset);
        offset += 4;
        writeU32(dataSize);
        // samples
        for (const s of this.samples) {
            buffer.writeInt16LE(s, offset);
            offset += 2;
        }
        fs.writeFileSync(filename, buffer);
    }

    play(filename) {
        const platform = process.platform;
        let cmd;
        if (platform === 'win32') {
            cmd = `start "" "${filename}"`;
        } else if (platform === 'darwin') {
            cmd = `afplay "${filename}"`;
        } else {
            // Linux
            const players = ['ffplay', 'aplay', 'mpg123'];
            // We'll try using the first available
            cmd = `ffplay "${filename}" || aplay "${filename}" || mpg123 "${filename}"`;
        }
        exec(cmd, (err) => {
            if (err) console.log('Playback error:', err.message);
        });
    }
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function interactive() {
    console.log('=== Waveform Audio Generator ===');
    const types = ['sine', 'square', 'sawtooth', 'triangle', 'noise'];
    let wtype = await ask(`Enter type (${types.join(', ')}, default: sine): `);
    wtype = wtype.trim() || 'sine';
    if (!types.includes(wtype)) wtype = 'sine';
    let freq = parseFloat(await ask('Enter frequency (Hz, default: 440): ') || '440');
    if (isNaN(freq) || freq <= 0) freq = 440;
    let amp = parseFloat(await ask('Enter amplitude (0.0-1.0, default: 0.8): ') || '0.8');
    if (isNaN(amp) || amp < 0 || amp > 1) amp = 0.8;
    let dur = parseFloat(await ask('Enter duration (seconds, default: 2.0): ') || '2.0');
    if (isNaN(dur) || dur <= 0) dur = 2.0;
    let rate = parseInt(await ask('Enter sample rate (Hz, default: 44100): ') || '44100');
    if (isNaN(rate) || rate <= 0) rate = 44100;
    let fname = await ask('Enter output filename (default: output.wav): ');
    fname = fname.trim() || 'output.wav';
    if (!fname.endsWith('.wav')) fname += '.wav';
    const gen = new WaveGenerator(freq, amp, dur, rate, wtype);
    gen.saveWAV(fname);
    console.log(`Saved to ${fname}`);
    const playAns = await ask('Play? (y/n, default: y): ');
    if (playAns.trim().toLowerCase() !== 'n') {
        gen.play(fname);
    }
    rl.close();
}

function cli() {
    const args = require('minimist')(process.argv.slice(2));
    const freq = args.f || args.freq || 440;
    const wtype = args.t || args.type || 'sine';
    const amp = args.a || args.amp || 0.8;
    const dur = args.d || args.dur || 2.0;
    const rate = args.r || args.rate || 44100;
    const output = args.o || args.output || 'output.wav';
    const noPlay = args['no-play'] || false;
    const gen = new WaveGenerator(parseFloat(freq), parseFloat(amp), parseFloat(dur), parseInt(rate), wtype);
    gen.saveWAV(output);
    console.log(`Saved to ${output}`);
    if (!noPlay) gen.play(output);
}

if (process.argv.length > 2) {
    cli();
} else {
    interactive();
}
